pkg/apis/autoscaling/validation/validation.go

```diff
@@ -127,3 +101,120 @@ func ValidateHorizontalPodAutoscalerStatusUpdate(newAutoscaler, oldAutoscaler *a
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.DesiredReplicas), field.NewPath("status", "desiredReplicasa"))...)
 	return allErrs
 }
+
+func validateMetrics(metrics []autoscaling.MetricSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for i, metricSpec := range metrics {
+		idxPath := fldPath.Index(i)
+		if targetErrs := validateMetricSpec(metricSpec, idxPath); len(targetErrs) > 0 {
+			allErrs = append(allErrs, targetErrs...)
+		}
+	}
+
+	return allErrs
+}
+
+var validMetricSourceTypes = sets.NewString(string(autoscaling.ObjectSourceType), string(autoscaling.PodsSourceType), string(autoscaling.ResourceSourceType))
+var validMetricSourceTypesList = validMetricSourceTypes.List()
+
+func validateMetricSpec(spec autoscaling.MetricSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(string(spec.Type)) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("type"), "must specify a metric source type"))
+	}
+
+	if !validMetricSourceTypes.Has(string(spec.Type)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("type"), spec.Type, validMetricSourceTypesList))
+	}
+
+	typesPresent := sets.NewString()
+	if spec.Object != nil {
+		typesPresent.Insert("object")
+		if typesPresent.Len() == 1 {
```

Comment:
I don't understand this if. It should be always true (we inserted "object" to an empty set `typesPresent`).